src/seqn/seqn.lisp

```diff
@@ -0,0 +1,139 @@
+  "Fills in extra inputs on the right with 0-oes"
+  (geb.utils:apply-n num (lambda (x) (append x (list 0))) seq))
+
+(defun 0-list (n)
+  (if (= n 1)
+      (list 0)
+      (cons 0 (0-list (- n 1)))))
+
+(defun prod-list (l1 l2)
+  "Takes two lists of same length and gives pointwise product
+where first element come from first list and second from second
+
+```lisp
+SEQN> (prod-list (list 1 2) (list 3 4))
+((1 3) (2 4))
+```"
+  (if (or (null l1)
+          (null l2))
+      nil
+      (let ((lh1 (length l1))
+            (lh2 (length l2)))
+        (if (= lh1 1)
+            (list (list (car l1) (car l2)))
+            (append (prod-list (butlast l1 1)
+                               (butlast l2 1))
+                    (list
+                     (list (nth (1- lh1) l1) (nth (1- lh2) l2))))))))
+
+(defun seq-max-fill (seq1 seq2)
+  "Takes two lists, makes them same length by adding 0es on the right
+where necessary and takes their pointwise product"
+  (let ((l1 (length seq1))
+        (l2 (length seq2)))
+    (if (< l1 l2)
+        (prod-list (fill-in (- l2 l1) seq1) seq2)
+        (prod-list seq1 (fill-in (- l1 l2) seq2)))))
+
+(defun max-list (lst)
+  (max (car lst) (car (cdr lst))))
+
+(defmethod seqwidth ((obj <substobj>))
+  (typecase-of substobj obj
+    (so0 (list 0))
+    (so1 (list 0))
+    (coprod (cons 1 (mapcar #'max-list
+                            (seq-max-fill
+                             (seqwidth
+                              (mcar obj))
+                             (seqwidth
+                              (mcadr obj))))))
+    (prod (append (seqwidth (mcar obj))
+                  (seqwidth (mcadr obj))))
+    (otherwise (geb.utils:subclass-responsibility obj))))
+
+(defun inj-coprod-parallel (
+                            obj     ;already transformed mcar or mcadr
+                            copr    ;already transformed coproduct
+                            )
+  "takes an A x B and gives an appropriate injection of
+(a1,...,an) into (max (a1, b1), ...., max(an, bn),...)
+i.e. where the maxes are being taken during the seqwidth operation
+without fulling in of the smaller object"
+  (if (= (length obj) (- (length copr) 1))
+      (drop-width obj  (cdr copr))
+      (composition (drop-width (append obj
+                                       (0-list (- (1- (length copr))
+                                                  (length obj))))
+                               (cdr copr))
+                   (inj-length-left obj (0-list (- (1- (length copr))
+                                                   (length obj)))))))
+
+(defmethod dom ((x <seqn>))
+  "Gives the domain of a morphism in SeqN.
+For a less formal desription consult the specs file"
+  (typecase-of seqn x
+    (composition      (dom (mcadr x)))
+    (fork-seq         (mcar x))
+    (parallel-seq     (append (dom (mcar x)) (dom (mcadr x))))
+    (id               (mcar x))
+    (drop-nil         (mcar x))
+    (remove-right     (append (mcar x) (list 0)))
```

Comment:
so many appens on stuff, maybe we should consider a different data structure at some point.